backend-rs/src/utils.rs

```diff
@@ -0,0 +1,425 @@
+            (
+                self.camera_mode.width as usize,
+                self.camera_mode.height as usize,
+            ),
+        ));
+        let bitstream = self.h264_encoder.encode(&yuv_buffer)?;
+
+        Ok(bitstream.to_vec())
+    }
+}
+
+// CameraMode(s) are used for controlling what resolution and frame rate the H264CameraReader operates at.
+#[derive(Debug, Clone, Copy)]
+pub struct CameraMode {
+    width: u32,
+    height: u32,
+    frame_interval: Fraction, // 1/30 is 30 fps, 1/15 is 15fps, etc.
+}
+
+impl ToString for CameraMode {
+    fn to_string(&self) -> String {
+        format!(
+            "{}x{} @{}fps",
+            self.width, self.height, self.frame_interval.denominator
+        )
+    }
+}
+
+impl CameraMode {
+    // Fetch all possible CameraMode(s) of the Device camera.
+    pub fn fetch_all(device: &Device) -> Result<Vec<CameraMode>, Box<dyn Error>> {
+        let mut configs: Vec<CameraMode> = Vec::new();
+
+        let discrete_iter = Capture::enum_framesizes(device, FourCC::new(b"MJPG"))?
+            .into_iter()
+            .flat_map(|framesize| framesize.size.to_discrete());
+
+        for discrete_frame_size in discrete_iter {
+            for interval in Capture::enum_frameintervals(
+                device,
+                FourCC::new(b"MJPG"),
+                discrete_frame_size.width,
+                discrete_frame_size.height,
+            )? {
+                let fraction = match interval.interval {
+                    FrameIntervalEnum::Discrete(fraction) => fraction,
+                    FrameIntervalEnum::Stepwise(Stepwise { min, .. }) => min,
+                };
+
+                configs.push(CameraMode {
+                    width: discrete_frame_size.width,
+                    height: discrete_frame_size.height,
+                    frame_interval: fraction,
+                });
+            }
+        }
+
+        Ok(configs)
+    }
+}
+
+// Handles communication between CameraThreadHandle(s) and WebRTC clients.
+pub struct WebcamManager {
+    camera_handles: Arc<Mutex<Vec<CameraThreadHandle>>>,
+    rtc_api: API,
+}
+
+impl WebcamManager {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        let mut media_engine = MediaEngine::default();
+        media_engine.register_default_codecs()?;
+
+        let mut registry = Registry::new();
+        registry =
+            interceptor_registry::register_default_interceptors(registry, &mut media_engine)?;
+
+        let rtc_api = APIBuilder::new()
+            .with_media_engine(media_engine)
+            .with_interceptor_registry(registry)
+            .build();
+
+        Ok(Self {
+            camera_handles: Arc::new(Mutex::new(Vec::new())),
+            rtc_api,
+        })
+    }
+
+    pub fn camera_handles(&self) -> Arc<Mutex<Vec<CameraThreadHandle>>> {
+        self.camera_handles.clone()
+    }
+
+    // IMPORTANT: Everything in here runs within the tokio runtime!
+    pub async fn add_client(
```

Comment:
I appreciate the notice above. However, the exact meaning of a "client" might be unclear far into the future. Please consider adding doc comments to this method.